docs/proposals/network-policy.md

```diff
@@ -0,0 +1,316 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NetworkPolicy
+
+## Abstract
+
+A proposal for implementing a new resource - NetworkPolicy - which
+will enable definition of ingress policies for selections of pods.
+
+The design for this proposal has been created by, and discussed
+extensively within the Kubernetes networking SIG.  It has been implemented
+and tested using Kubernetes API extensions by various networking solutions already.
+
+In this design, users can create various NetworkPolicy objects which select groups of pods and
+define how those pods should be allowed to communicate with each other.  The
+implementation of that policy at the network layer is left up to the
+chosen networking solution.
+
+> Note that this proposal does not yet include egress / cidr-based policy, which is still actively undergoing discussion in the SIG. These are expected to augment this proposal in a backwards compatible way.
+
+## Implementation
+
+The implmentation in Kubernetes consists of:
+- A v1beta1 NetworkPolicy API object
+- A structure on the `Namespace` object to control policy, to be developed as an annotation for now.
+
+### Namespace changes
+
+The following objects will be defined on a Namespace Spec.
+>NOTE: In v1beta1 these objects will be implemented as an annotation.
+
+```go
+type IngressIsolationPolicy string
+
+const (
+	// Deny all ingress traffic to pods in this namespace.
+	DefaultDeny IngressIsolationPolicy = "DefaultDeny"
+) 
+
+// Standard NamespaceSpec object, modified to include a new
+// NamespaceNetworkPolicy field.
+type NamespaceSpec struct {
+	// This is a pointer so that it can be left undefined.
+	NetworkPolicy *NamespaceNetworkPolicy `json:"networkPolicy,omitempty"`
```

Comment:
We've begun discussing implementing group policy on namespaces via label selection (for things like bulk quota and bulk limit ranges).  Has that been discussed (how to define network policy for a large swath of the cluster)?  Or started as a concrete reference on the namespace because it's easy to reason about and doesn't require a more complex story?